src/SimpleMutex.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Thesis\MessageBus\Transport\Amqp;
+
+use Amp\Cancellation;
+use Amp\Future;
+use function Amp\async;
+
+/**
+ * @api
+ * @template T
+ */
+final class SimpleMutex
+{
+    /**
+     * @var ?Future<T>
+     */
+    private ?Future $future = null;
+
+    /**
+     * @var ?T
+     */
+    private mixed $value = null;
+
+    /**
+     * @var \Closure(T): bool
+     */
+    private readonly \Closure $isHit;
```

Comment:
A possible better name could be `isAlive`, since the value will most often be hit, but may no longer be valid.